### ISSUE TITLE
Enable worker-level profiling of Dataflow Jobs 

### ISSUE THREAD

**Issue by bjchambers** [GoogleCloudPlatform/DataflowJavaSDK#72]
For both working on the SDK and building Dataflow pipelines, it would be useful if there was an easy way to get profiles from the execution of code on the workers.

**Comment by bjchambers** [comment]
We’re working on a better experience for profiling but there is rudimentary support for profiling available in the SDK.

# What you’ll need
1. An installation of [pprof](https://github.com/google/pprof) 
2. An installation of [graphviz](http://www.graphviz.org) if you’d like to visualize profile information.

# How to get profiles

1. **Run your pipeline** with `--enableProfilingAgent`. This will write profiles locally in the worker VM to either `/var/log/dataflow/java-batch/cprof` for batch jobs or `/var/log/dataflow/java-streaming/cprof` for streaming jobs.

2. **Retrieve the profiles** from the workers using `gcloud compute copy-files`. For example:

        gcloud compute copy-files --project=[PROJECT] --zone=[ZONE] \   
          [VM_INSTANCE]:/var/log/dataflow/java-['batch' or 'streaming']/cprof \
          [LOCAL_DIR]

3. **View the profiles** using `pprof.  Run `pprof [LOCAL_DIR]/*cpu*.gz` for CPU profiles (or `*wall*.gz` for wall-time profiles). From here you can run `graphviz` to render a calltree, or `text` or `tree` for text-based reports. See the [pprof docs](https://github.com/google/pprof/blob/master/doc/pprof.md) and `pprof --help` for more ways to interact with the profiles.

Hope that helps!

# Notes and Caveats
- The profiles will be 10 second samples from every 60 seconds of execution.
- For a batch job the VM instances are normally torn down after the job completes, making it a race to get the profiles. If you run with `--teardownPolicy=TEARDOWN_NEVER` the VMs will be left around forever. You will then need to delete the VMs manually using gcloud or the GCE console.
- Multiple `ParDo` steps may execute together. When this happens, the call to `output()` in the first step will include the time to execute the later steps. As a result, the inclusive time for these steps will be inflated.
- If you want the profiles to include information about JNI calls make sure to have any relevant binaries/object files in the directory you run pprof from.

**Comment by bfabry** [comment]
> If you want the profiles to include information about JNI calls make sure to have any relevant binaries/object files in the directory you run pprof from.

Is there any documentation on how to get the binaries used by dataflow to do this?

**Comment by peay** [comment]
Is there similar support for Beam's Dataflow runner?

**Comment by swegner** [comment]
Yes, in Apache Beam profiling support is now enabled via `--saveProfilesToGcs=<gs://...>`, defined inside [`DataflowProfilingOptions`](https://github.com/apache/beam/blob/v2.4.0/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowProfilingOptions.java).

**Comment by bvolpato** [comment]
I couldn't get it to work.

Even though I am sending:

```
  saveProfilesToGcs: gs://labs1-carol-internal/profiler
  profilingAgentConfiguration: {APICurated=true}
```

Through Java code:
```
        DataflowProfilingOptions profilingOptions = dataflowPipelineOptions.as(DataflowProfilingOptions.class);
        profilingOptions.setSaveProfilesToGcs("gs://" + PipelineHelper.getBucketName(bucket) + "/profiler");

        DataflowProfilingAgentConfiguration agent = new DataflowProfilingAgentConfiguration();
        agent.put("APICurated", true);
        
        profilingOptions.setProfilingAgentConfiguration(agent);
        
```

I don't get any files in the profiler, and this message is printed on Stackdriver:
`Profiling Agent not found. Profiles will not be available from this worker.`

Any ideas?

**Comment by lukecwik** [comment]
Which version of the SDK are you using?

Have you tried contacting Google Cloud support and share some job ids with them?

**Comment by bvolpato** [comment]
@lukecwik I tried with both 2.13.0 and 2.14.0. Will try to contact their support, thanks!

**Comment by bvolpato** [comment]
Support could not help with this and still didn't find a way to get profilers.

> On the other hand, I would like to mention that Profiles in Dataflow don't have a Service Level Agreement (SLA) since this is an experimental Alpha feature, and is not recommended for production use cases, as mentioned in [2]. 
> [2] https://cloud.google.com/products/#product-launch-stages

**Comment by bvolpato** [comment]
For those who are wondering, the Profiler does not get populated (and profile files are not saved on GCS, either) if you set both properties at the same time (`APICurated=true` and `saveProfilesToGcs={path}`).

I removed the `saveProfilesToGcs` and now [profiler](https://console.cloud.google.com/profiler/) works fine for me.
